pypeit/images/rawimage.py

```diff
@@ -424,38 +726,100 @@ def subtract_pattern(self):
+            force (:obj:`bool`, optional):
+                Force the image to be trimmed, even if the step log
+                (:attr:`steps`) indicates that it already has been.
         """
         step = inspect.stack()[0][3]
-        # Check input image matches the original
         if self.rawimage.shape is not None and self.image.shape != self.rawimage.shape:
-            msgs.warn("Image shape does not match original.  Returning current image")
-            return self.image.copy()
-        # Check if already trimmed
-        if self.steps[step] and (not force):
-            msgs.warn("Image was already trimmed.  Returning current image")
-            return self.image
-        # Do it
+            # Image *must* have been trimmed already because shape does not
+            # match raw image
+            self.steps[step] = True
+            msgs.warn('Image shape does not match original.')
+            return
+        if self.steps[step] and not force:
+            # Already trimmed
+            msgs.warn('Image was already trimmed.')
+            return
         self.image = procimg.trim_frame(self.image, self.datasec_img < 1)
+        if self.var is not None:
+            self.var = procimg.trim_frame(self.var, self.datasec_img < 1)
+        if self.rn2img is not None:
+            self.rn2img = procimg.trim_frame(self.rn2img, self.datasec_img < 1)
+        # NOTE: This must be done last because the untrimmed image is used for
+        # deciding what to trim!
         self.datasec_img = procimg.trim_frame(self.datasec_img, self.datasec_img < 1)
-        #
         self.steps[step] = True
 
+    def add_shot_noise(self):
+        r"""
+        Use the image to calculate and add shot noise to the error budget.
+
+        The method assumes the image is bias- and dark-subtracted, and that the
+        units are electrons/counts.  The calculation of the shot noise includes
+        a term that adjusts the variance at low count levels to account for the
+        Gaussian approximation of a Poisson distribution throughout the rest of
+        the code base (*need a reference for this*).  If availabile, the
+        calculation also includes shot noise from the dark current, using the
+        exposure time and the dark current (e-/s) provided by the spectrograph
+        detector (:class:`~pypeit.images.detector_container.DetectorContainer`)
+        object.  In detail, the added shot noise variance is:
+
+        .. math::
+
+            V = | C - \sqrt(2 V_{\rm rn}) | + C_{\rm dark} t_{\rm exp}
+
+        where :math:`V_{\rm rn}` is the readnoise variance (see
+        :func:`~pypeit.core.procimg.rn2_frame`), :math:`C` is the number of
+        bias- and dark-subtracted counts, :math:`C_{\rm dark}` is the number of
+        dark counts per second, and :math:`t_{\rm exp}` is the exposure time
+        (:attr:`exptime`) in seconds.
+
+        .. note::
+
+            The variance attribute (:attr:`var`) is edited in-place.
+
+        """
```

Comment:
Can we wrap a core function here. I think that would more consistent with the pypeIt programming approach and I also need to update the noise model using sky and object models in local_skysub_extract and it would be best if were were calling the same core method there.

Reply:
Yes, I now wrap a core function.